### PR TITLE
Bump CAPI to v1.8.1

### LINKED ISSUE #4991: CAPI v1.8.0-beta.0 has been released and is ready for testing
CAPI v1.8.0-beta.0 has been released and is ready for testing.
Looking forward to your feedback before v1.8.0 release!

## For quick reference

- [CAPI v1.8.0-beta.0 release notes](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.8.0-beta.0)
- [Shortcut to CAPI git issues](https://github.com/kubernetes-sigs/cluster-api/issues)

## Following are the planned dates for the upcoming releases

CAPI v1.8.0 will be released on **Monday, 12th August 2024**.

More details of the upcoming schedule can be seen at [CAPI v1.8.0 release timeline](https://github.com/kubernetes-sigs/cluster-api/tree/main/docs/release/releases/release-1.8.md#timeline).

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Updates CAPI to [v1.8.1](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.8.1) and all that entails.

**Which issue(s) this PR fixes**:

Fixes #4991

**Special notes for your reviewer**:

See the [migration guide](https://main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.7-to-v1.8) for more information.

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Bump CAPI to v1.8.1
```

### CODE REVIEW

**Comment by mboersma** [comment]
/retitle [WIP] Bump CAPI to v1.8.0-rc.0

Vendoring changes here are giving me fits. I could use another pair of eyes, or ideas if anyone has any.

**Comment by troy0820** [comment]
> Vendoring changes here are giving me fits. I could use another pair of eyes, or ideas if anyone has any.

It looks like you have an issue here:
`sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.17.3`

That's one of the issues at least.

**Comment by mboersma** [comment]
It is one of the issues indeed, thanks @troy0820.

I started removing those `replace` lines locally, but it's a slippery slope and obscure compilation errors continue to accumulate. But that one in particular does need to go, we can start there.

**Comment by jackfrancis** [comment]
@mboersma just curious, is this PR making this more difficult? https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4973

**Comment by troy0820** [comment]
It's your replaces at the end of the go.mod.  The older k8s.io/* deps are clashing with what is imported by controller-runtime and cluster-api.  

>It is one of the issues indeed, thanks @troy0820.
I started removing those replace lines locally, but it's a slippery slope and obscure compilation errors continue to accumulate. But that one in particular does need to go, we can start there.

Where I am after commenting out all the old replaces was issues with:
image

in `controllers/azurejson_machinetemplate_controller.go` I can provide diff or make a PR to this branch if I do it fast enough.

**Comment by mboersma** [comment]
> is this PR making this more difficult? #4973

Somewhat, but there's a clear comment that those `replace`s may need to go away with this change. I was narrowing in on which ones need to remain, but got muddled every time I got close. @nojnhuh may have found the recipe, I merged his changes just now.

**Comment by troy0820** [comment]
@nojnhuh  was faster than I was, but if you want to keep parity with upstream capi, this may be the release you ditch the replaces because I believe cr is running k8s.io/* at v0.30.0 starting in v0.18.0 which this uses.  some apis are gone from that (none that we need to worry about) but things changing in controller-runtime/capi may force your hand.

**Comment by nojnhuh** [comment]
The AKS failure seems 100% consistent when I run the tests locally, but I can't reproduce with the same template using Tilt. CAPZ gets stuck in a panic loop because the `spec.controlPlaneRef` on the Cluster isn't set after it's created from a ClusterClass. Only the AKS ClusterClass test seems affected, not the self-managed ClusterClass test.

**Comment by nojnhuh** [comment]
This may fix the AKS clusterclass test somehow? #5058

**Comment by nojnhuh** [comment]
> Code: RoleAssignmentLimitExceeded

commenced role assignment purge

**Comment by nojnhuh** [comment]
> The AKS failure seems 100% consistent when I run the tests locally, but I can't reproduce with the same template using Tilt. CAPZ gets stuck in a panic loop because the `spec.controlPlaneRef` on the Cluster isn't set after it's created from a ClusterClass. Only the AKS ClusterClass test seems affected, not the self-managed ClusterClass test.

> This may fix the AKS clusterclass test somehow? #5058

@mboersma ^I think I have a handle on the failing AKS test

**Comment by mboersma** [comment]
This is finally passing tests and ready for review.

**Review by nojnhuh — APPROVED** [review]
/lgtm
/assign @willie-yao

File: test/e2e/config/azure-dev.yaml [hunk 1]
```diff
@@ -224,7 +224,7 @@ variables:
   SECURITY_SCAN_CONTAINER: "${SECURITY_SCAN_CONTAINER:-quay.io/armosec/kubescape:v2.0.167}"
   AZURE_CNI_V1_MANIFEST_PATH: "${PWD}/templates/addons/azure-cni-v1.yaml"
   OLD_CAPI_UPGRADE_VERSION: "v1.6.8"
```
Comment (willie-yao) [inline test/e2e/config/azure-dev.yaml:226]: Would we wanna bump this as well to 1.7.5?

**Review by nojnhuh — APPROVED** [review]
/lgtm
/approve

**Comment by willie-yao** [comment]
> Error: failed to get provider components for the "cluster-api:v1.7.5" provider: failed to read "components.yaml" from provider's repository "cluster-api": failed to read file "/logs/artifacts/repository/cluster-api/v1.7.5/components.yaml" from local release v1.7.5: stat /logs/artifacts/repository/cluster-api/v1.7.5/components.yaml: no such file or directory

I think you might need to change some of the paths in azure-dev.yaml back to 1.7.5

**Pull request merged**